pkg/flavor/kubernetes/multiwriter.go

```diff
@@ -0,0 +1,673 @@
+// Copyright 2019 Hewlett Packard Enterprise Development LP
+
+package kubernetes
+
+import (
+	"fmt"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	log "github.com/hpe-storage/common-host-libs/logger"
+	apps_v1 "k8s.io/api/apps/v1"
+	core_v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"os"
+	"strings"
+	"time"
+)
+
+const (
+	serviceNamePrefix    = "hpe-multiwriter-svc-"
+	deploymentNamePrefix = "hpe-multiwriter-"
+	pvcNamePrefix        = "hpe-multiwriter-pvc"
+	multiWriterNamespace = "kube-system"
+	nfsImage             = "quay.io/luis_pabon0/ganesha" // TODO evaluate latest nfs-ganesha image
+	deletionInterval     = 30                            // 60s with sleep interval of 2s
+	deletionDelay        = 2 * time.Second
+	creationInterval     = 60 // 120s with sleep interval of 2s
+	creationDelay        = 2 * time.Second
+	// defaultVolumeSize is the implementation-specific default value in bytes
+	defaultVolumeSize = 10 * 1024 * 1024 * 1024 // 10 GiB
+	//defaultExportPath = "/exports"
+)
+
+// CreateMultiWriterVolume creates multi-writer volume abstracting underlying nfs pvc, deployment and service
+func (flavor *Flavor) CreateMultiWriterVolume(request *csi.CreateVolumeRequest) (multiWriterVolume *csi.Volume, rollback bool, err error) {
+	log.Tracef(">>>>> CreateMultiWriterVolume with %s", request.Name)
+	defer log.Tracef("<<<<< CreateMultiWriterVolume")
+
+	// get pvc from request
+	claim, err := flavor.getClaimFromClaimName(request.Name)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// clone pvc and modify to RWO mode and destroyOnDelete
+	claimClone, err := flavor.makeNFSPvc(claim)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// create pvc
+	newClaim, err := flavor.CreateNFSPVC(claimClone)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// create deployment with name hpe-multiwriter-<originalclaim-uid>
+	deploymentName := fmt.Sprintf("%s%s", deploymentNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.CreateNFSDeployment(deploymentName, newClaim.ObjectMeta.Name)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// create service with name hpe-multiwriter-svc-<originalclaim-uid>
+	serviceName := fmt.Sprintf("%s%s", serviceNamePrefix, claim.ObjectMeta.UID)
+	err = flavor.CreateNFSService(serviceName)
+	if err != nil {
+		return nil, true, err
+	}
+
+	// Get capacity requested
+	reqVolumeSize := int64(defaultVolumeSize) // Default value
+	if request.GetCapacityRange() != nil {
+		reqVolumeSize = request.GetCapacityRange().GetRequiredBytes()
+	}
+
+	// Return newly created volume
+	return &csi.Volume{
+		VolumeId:      fmt.Sprintf("%s", claim.ObjectMeta.UID),
+		CapacityBytes: reqVolumeSize,
+		VolumeContext: make(map[string]string),
```

Comment:
We can't leave this empty.  This needs to be populated with the create options so the PV can describe its details.

Reply:
got it..